examples/on-conflict-clause/on_conflict_clause.bal

```diff
@@ -7,32 +7,53 @@ type Student record {|
 |};
 
 public function main() {
-    error onConflictError = error("Key Conflict", message = "record with same key exists.");
-
     table<Student> students = table [
             {id: 1, name: "John", score: 100},
             {id: 2, name: "Jane", score: 150},
             {id: 1, name: "John", score: 200}
         ];
 
-    // The result of the following will be an error since the key `John` is duplicated.
+    // The result of the following will be an error since the key `John` is duplicated
+    // and `isSafeReplace()` function returns an error for the name `John`.
     map<int>|error studentScores = map from var {name, score} in students
                                    select [name, score]
-                                   on conflict onConflictError;
+                                   on conflict isSafeReplace(name);
 
     io:println(studentScores);
 
-    // The value `100` of the key `John` will be replaced by the second occurrence value `200`.
-    map<int> lastRoundScore = map from var student in students
-                              select [student.name, student.score]
-                              on conflict ();
-
-    io:println(lastRoundScore);
-
-    // The result of the following will be an error since the key `1` is duplicated.
+    // The result of the following will be an error since the key `1` is duplicated
+    // and `isSafeReplace()` function returns an error for the name `John`.
     table<Student> key(id)|error studentScoresTable = table key(id) from var student in students
                                                       select student
-                                                      on conflict onConflictError;
+                                                      on conflict isSafeReplace(student.name);
 
     io:println(studentScoresTable);
+
+    table<Student> students2 = table [
+            {id: 1, name: "John", score: 100},
+            {id: 2, name: "Mike", score: 150},
+            {id: 2, name: "Mike", score: 200}
+        ];
+
+    // The value `100` of the key `Mike` will be replaced by the second occurrence value `200`
+    // since `isSafeReplace()` function returns `null` for the name 'Mike`.
+    map<int>|error lastRoundScore = map from var student in students2
+                                    select [student.name, student.score]
+                                    on conflict isSafeReplace(student.name);
+
+    io:println(lastRoundScore);
+
+    // The return value will be just the construct type if the `on conflict` clause
+    // always returns `()`.
+    map<int> lastRoundScore2 = map from var student in students2
+                               select [student.name, student.score]
+                               on conflict ();
+
+    io:println(lastRoundScore2);
+}
+
+function isSafeReplace(string name) returns error? {
+    if name == "John" {
```

Comment:
IMO, a check like this still doesn't make it like a real world example? What if we just have two examples, one where we just use an error constructor in the on conflict clause to return an error (or call a function that logs at error level and returns an error) and the other call a function that just logs a warning saying the existing value is replaced on conflict and returns nil?

Reply:
IMO, we should demonstrate error handling given a frame variable. Param `name` in this case. Static error returning and logging are not ideal. Using a frame variable in error handling is the purpose of creating this issue earlier. Hence, I think the existing function is simpler and demonstrates the purpose

Reply:
We can just include `name` in the error/warning log or the error value? 

BBEs shouldn't be too complicated. We just need to make sure the idea is conveyed in a meaningful example.

Reply:
Including the `name` in the log and always returning an error is not the ideal sample right? We must convey to the user that he can do some logic with the frame variables and decide whether to replace the value or give an error. Also, I believe the if condition complicates the BBE since it is very straightforward.

Reply:
> We must convey to the user that he can do some logic with the frame variables and decide whether to replace the value or give an error.

Isn't that conveyed via having a function being called with the variable as an argument? I think it's self-explanatory. 

> Including the `name` in the log and always returning an error is not the ideal sample right?

What we want to do IMO is explain that you can return error or nil and how the behaviour changes. 

I think the common case is that for all on conflict scenarios you either return error or you return nil. Rather than returning error in some cases and nil in others. 

If we really need an error? example, still checking for an arbitrary name like we've done here isn't real-world-like IMO. A better check would be to check for restricted names like `admin`, which still isn't ideal. Can we rethink the example?

Reply:
+1 to change the name to `admin`. Why would you think it still isn't ideal? It conveys to the user that the function returns errors for some special cases depending on the frame variable.

Reply:
Because it's not something you'd even expect being in the collection to start with (i.e., not accessible to the end user).

Reply:
The idea that we want to demonstrate is to show the user that he can use the frame variable and depending on that he can do some logic to decide whether to replace the value or give the error. Since, this sample shows that exactly, IMO this sample is enough without complications. WDYT?

Reply:
But why are we arbitrarily checking for John? That's my main concern. We're saying we're trying to make it more like a real-world example, but we aren't entirely.

Reply:
The idea is there, right? Shall we change the name to `admin` as you suggested? We can change the logic to something more natural but it will complicate the logic. IMO even though the sample is not entirely a real-world case, it demonstrates the real-world purpose or the idea without complications. WDYT?

Reply:
I'm still not convinced this is the correct example. We're assuming the first entry in the table is the correct entry. 

@gimantha any thoughts on this? Since we're going back and forth. :)